File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;
class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> max = new PriorityQueue<>((o1,o2)-> o2.compareTo(o1));
        PriorityQueue<Integer> min = new PriorityQueue<>();
        for (String s: operations){
            String[] ss = s.split(" ");
            if (ss[0].equals("I")){
                min.offer(Integer.parseInt(ss[1]));
                max.offer(Integer.parseInt(ss[1]));
            }
            else {
                if (max.size()>0){
                    if (Integer.parseInt(ss[1])>0){
                        min.remove(max.poll());
                    }
                    else {
                        max.remove(min.poll());
                    }
                }
            }
        }
        int[] answer = new int[2];
        if (max.size()>0) {
            answer[0] = max.poll();
            answer[1] = min.poll();
        }
        return answer;
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
def solution(operations):
    min=[]
    max=[]
    visit = dict()
    for ind,op in enumerate(operations):
        command, value = op.split(" ")
        value = int(value)
        if command == 'I':
            heapq.heappush(min,(value,ind))
            heapq.heappush(max,(-value,ind))
            visit[ind]=1
        elif value == 1:
            while max and visit.get(max[0][1])==0:
                heapq.heappop(max)
            if max:
                visit[heapq.heappop(max)[1]]=0
        elif value ==-1:
            while min and visit.get(min[0][1])==0:
                heapq.heappop(min)
            if min:
                visit[heapq.heappop(min)[1]]=0
    while max and visit.get(max[0][1])==0:
                heapq.heappop(max)
    while min and visit.get(min[0][1])==0:
                heapq.heappop(min)
    return [0,0] if not min or not max else [-heapq.heappop(max)[0], heapq.heappop(min)[0]]